federation/planner.go

```diff
@@ -116,6 +139,30 @@ func printSelections(selectionSet *graphql.SelectionSet) {
 	}
 }
 
+// maybeCustomService adds the input selection to either localSelections or
+// selectionByService if there is a serviceSelector in the planner and the user
+// has custom service setting for the selection. It returns a boolean to indicates
+// whether there is a custom service for the selection.
+func (e *Planner) maybeCustomService(
+	typeName, currentService string,
+	selection *graphql.Selection,
+	localSelections []*graphql.Selection,
+	selectionsByService map[string][]*graphql.Selection,
+) (bool, []*graphql.Selection, map[string][]*graphql.Selection) {
+	if e.serviceSelector == nil {
+		return false, localSelections, selectionsByService
+	}
+	customService := e.serviceSelector(typeName, selection.Name)
+	if customService == "" {
+		return false, localSelections, selectionsByService
+	}
+	if customService == currentService {
```

Comment:
Can we make it such that `maybeCustomService` doesn't actually change the local selections or selectionsByService and returns the custom service. This makes it to understand since `maybeCustomService` isn't mutating any of the local fields and`planObject` can be responsible for all the changes to local selections or selections by service.

Reply:
Sure, changed!

Reply:
Actually I changed it back, the reason is the check on line 221 will fail if we don't change the local selections return a false while we are migrating the field.

Reply:
I meant changing it such that the function we call tells us what service to execute the query on and does all th evalidations, but doesn't actually change the list of selections. This makes the code a lot easier to follow in terms of understanding where the list of selections are changing. I put in an example below
```
func (e *Planner) selectService(
	typeName, currentService string,
	selection *graphql.Selection,
	field *graphql.Field,
	fieldInfo *FieldInfo,
) (error, string) {
	customService := ""
	if e.serviceSelector != nil {
		customService = e.serviceSelector(typeName, selection.Name)
	}
	if customService == "" {
		if len(field.FederatedKey) == 0 && len(fieldInfo.Services) != 1 {
			return oops.Errorf("field %s is not in serviceSelector and has %d associate services, current service %s", selection.Name, len(fieldInfo.Services), currentService), ""
		}
		if fieldInfo.Services[currentService] {
			return nil, currentService
		}
		for service, hasField := range fieldInfo.Services {
			if hasField {
				return nil, service
			}
		}
		return oops.Errorf("Field is not on multiple services"), ""
	}
	if len(field.FederatedKey) > 0 {
		return oops.Errorf("It is not allowed to select service for non field func %s for type %s", selection.Name, typeName), ""
	}
	if _, ok := fieldInfo.Services[customService]; !ok {
		return oops.Errorf("no service %s", customService), ""
	}
	return nil, customService
}

... 

err, customService := e.selectService(typ.Name, service, selection, field, fieldInfo)
		if err != nil {
			return nil, oops.Wrapf(err, "selecting service")
		}
		if customService == service {
			localSelections = append(localSelections, selection)
		} else {
			selectionsByService[customService] = append(selectionsByService[customService], selection)
		}
```

Reply:
Ah, I see, changed according to the comment!